clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -844,7 +847,18 @@ const FieldDecl *CodeGenFunction::FindFlexibleArrayMemberField(
       if (const FieldDecl *Field =
```

Comment:
FieldNo and Layout are referring to fields of "RD"; the "Field" found in the recursive visit is a member of Record (or some subobject of Record).  So the code is doing math on completely unrelated offsets.  Checking getFieldCount is just masking the issue.

Reply:
Maybe instead of looking for RecordDecls, this code should be looking for fields where the type of the field is an anonymous struct/union.

Reply:
> FieldNo and Layout are referring to fields of "RD"; the "Field" found in the recursive visit is a member of Record (or some subobject of Record). So the code is doing math on completely unrelated offsets. Checking getFieldCount is just masking the issue.

That's not the issue. What's happening is when an inner struct is declared/defined, we recurse within it to try to find the `Field` offset. If we do find it, then `Offset` has the offset value within `Record`. At this point, what we need is the offset up to the `RecordDecl`, but since those may or may not have a field number associated with them, we use the last `FieldNo` to get that offset.

A bit clearer:

```
struct foo {        /* <- Passed in as RD */
  struct bar {      /* <- Not a field, so FieldNo isn't incremented. Recurse on struct bar */
    int array[];    /* <- FAM found at offset 0 of struct bar */
  };
                    /* <- Returning with the array FieldDecl, we want to add on any
                          offset associated with the placement of the struct bar
                          definition, but there are no FieldDecls, and so we can't
                          call 'Layout.getFieldOffset()' */
};
```

This has obvious issues with virtual classes and the like, which is why C++ doesn't officially support FAMs (I believe it's an extension).

To be fair, I had the same question you had. I should document this better.

> Maybe instead of looking for RecordDecls, this code should be looking for fields where the type of the field is an anonymous struct/union.

We want to look into all inner structs to find a FAM that may be lurking deep down within the bowels of the struct, which may involve non-anonymous structs.

Reply:
I did several tests, and it looks like if there's an inner struct that's not accessible, that doesn't affect the offsets of fields outside of that inner struct.

```c
struct foo {
  struct inner_1 {
    /* fields */
  };
  struct inner_2 {
    int a;               /* __builtin_offsetof is 0 */
  };
  int b;                 /* __builtin_offsetof is 0 */
};
```

@efriedma-quic Do you have any thoughts?

Reply:
>  if there's an inner struct that's not accessible, that doesn't affect the offsets of fields outside of that inner struct.

Yes, the only thing that's relevant to offsets in a struct is the fields.  A struct declaration inside another struct declaration has exactly the same semantics as a struct definition anywhere else (unless it's an anonymous struct/union).

----

Maybe also try the following testcase.

```
struct foo {
  int x,y,z;
  struct bar {
    int count;
    int array[] __attribute__((counted_by(count)));
  };
};
void init(void * __attribute__((pass_dynamic_object_size(0))));
void test1(struct bar *p) {
  init(p->array);
}
```